docs(apphub): add App Hub CI guide (#31)
dhis2/developer-portal: docs/guides/publish-apphub.md

---
id: publish-apphub
title: How to setup continuous delivery to the App Hub
---

Once you have [submitted](/docs/guides/submit-apphub) your app and it's been reviewed by the DHIS2 Core Team, your app is live on the App Hub and can be installed directly from other DHIS2 instances through the App Management App. However, it is annoying and time consuming to upload new app versions manually, and difficult to keep it in sync with your development process. This guide will walk you through how to setup continuous delivery to the App Hub, so new versions are automatically built and pushed to the App Hub.

> Note that this guide assumes that your app built using the [DHIS2 Application Platform](https://platform.dhis2.nu/#/). See the [tutorial](/docs/tutorials/setup-env) for how to build your first DHIS2 app.

> > We will use [Github Actions](https://docs.github.com/en/actions) to power the continuous integration process. It is recommended that you to familiar yourself with the Github Actions' documentation, but it's not necessary for completion of this tutorial.

Reviewer: ```suggestion
> We will use [GitHub Actions](https://docs.github.com/en/actions) to power the continuous integration process. It is recommended that you familiarize yourself with the GitHub Actions' documentation, but it's not necessary for the completion of this tutorial. 
```

#### 1. Sign in to the [DHIS2 App Hub](https://apps.dhis2.org/)

Click the profile-icon on the top right and click "Your API Keys".

#### 2. Generate your API key

Click "Generate API key". Write this key down, you will not be able to see it again.

#### 3. Setting the repository secret

> Navigate to the Github repository of your app. Click "Settings" -> "Secrets". Click "New repostory secret". Use `D2_APP_HUB_API_KEY` as the name and paste the API-key in the previous step as the value.

Reviewer: ```suggestion
Navigate to the GitHub repository of your app. Click "Settings" -> "Secrets". Click "New repository secret". Use `D2_APP_HUB_API_KEY` as the name and paste the API-key in the previous step as the value.
```

#### 4. Setting the app id

The App Hub needs to know which app your app-version belongs to. This is done using the ID-field in the `d2.config.js`.

Navigate to the App Hub and find your app in the list. Click on the app, and copy the `id` from the browser URL. The ID should look like `b783bf32-0cc2-4d31-aadc-22d4c4807c30`.

Set the `id` field in the `d2.config.js` in your app to the id of the app on the App Hub.

The `minDHIS2Version` is compulsory as well, so set the minimum DHIS2 version that your app is compatible with.

> Double check that all the information is correct, the upload will fail if there's a mismatch betweent he App Hub information and the config-information.

Reviewer: ```suggestion
Double check that all the information is correct, the upload will fail if there's a mismatch between the App Hub information and the config-information.
```

```
// d2.config.js

 const config = {
    id: 'b783bf32-0cc2-4d31-aadc-22d4c4807c30',
    title: 'Simple Example App', // this should match the name of your app on the App Hub
    type: 'app',
    entryPoints: {
        app: './src/App.js',
    },
    minDHIS2Version: '2.34',
}
module.exports = config
```

> #### 4. Adding Github Action Workflow

Reviewer: ```suggestion
#### 4. Adding GitHub Action Workflow
```

> Here's an exmaple of a minimal workflow that you can use. This will build the app and publish it to the App Hub whenever a release is created on GitHub.

Reviewer: ```suggestion
Here's an example of a minimal workflow that you can use. This will build the app and publish it to the App Hub whenever a release is created on GitHub.
```

Copy this file to `.github/workflows/apphub-release.yml` in your repository. Remember to push the file. You may also use the GitHub interface for this (Actions -> New Workflow).

> ```

Reviewer: ```suggestion
```yml
```

  workflow_dispatch:

  build:
    # The type of runner that the job will run on
    runs-on: ubuntu-latest

    # Steps represent a sequence of tasks that will be executed as part of the job
    steps:
      # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2
        with:
          node-version: 12.x
      - uses: c-hive/gha-yarn-cache@v1

      - name: Install JS dependencies
        run: yarn install

      - name: Build
        run: yarn build

      - name: Release to DHIS2 App Hub
        run: yarn run d2-app-scripts publish
```

#### 5. Prepare a release

Run `yarn version --patch`. This will bump the version of your app, and your app should be ready for publishing. Push this to the remote: `git push origin main --follow-tags`.

#### 6. Create a release

Navigate to the app repository on GitHub and Click `Releases` on the Section on the right.

Click `draft a new release`. Use the tag-version that was created in the previous step (eg `v1.1.5`).
Optionally type in a release title and description and click `Publish release`.

> The GitHub action should be triggered and after a couple of minutes your new version should be pushed to the App Hub!

Reviewer: ```suggestion
The GitHub action should be triggered and after a couple of minutes your new version should be pushed to the App Hub! 🎉 
```

---
id: publish-apphub
title: How to setup continous delivery to the App Hub

---

Once you have [submitted](/docs/guides/submit-apphub) your app and it's been reviewed by the DHIS2 Core Team, your app is live on the App Hub and can be installed directly from other DHIS2 instances through the App Management App. However, it is annoying and time consuming to upload new app versions manually, and difficult to keep it in sync with your development process. This guide will walk you through how to setup continous delivery to the App Hub, so new versions are automatically built and pushed to the App Hub. 

> Note that this guide assumes that your app built using the [DHIS2 Application Platform](https://platform.dhis2.nu/#/). See the [tutorial](/docs/tutorials/setup-env) for how to build your first DHIS2 app.

> > We will use [Github Actions](https://docs.github.com/en/actions) to power the continous integration process. It is recommended that you to familiar yourself with the Github Actions' documentation, but it's not necessary for completion of this tutorial. 

Reviewer: ```suggestion
> We will use [GitHub Actions](https://docs.github.com/en/actions) to power the continuous integration process. It is recommended that you familiarize yourself with the GitHub Actions' documentation, but it's not necessary for the completion of this tutorial.
```

#### 1. Sign in to the [DHIS2 App Hub](https://apps.dhis2.org/)

Click the profile-icon on the top right and click "Your API Keys". 

#### 2. Generate your API key

Click "Generate API key". Write this key down, you will not be able to see it again.

#### 3. Setting the repository secret

> Navigate to the Github repository of your app. Click "Settings" -> "Secrets". Click "New repostory secret". Use `D2_APP_HUB_API_KEY` as the name and paste the API-key in the previous step as the value. 

Reviewer: ```suggestion
Navigate to the GitHub repository of your app. Click "Settings" -> "Secrets". Click "New repository secret". Use `D2_APP_HUB_API_KEY` as the name and paste the API-key in the previous step as the value.
```

#### 4. Setting the AppId

When uploading a new version to the App Hub, it needs to know which app to upload to. This is done using the ID-field in the `d2.config.js`.

Navigate to the App Hub and find your app in the list. Click on the app, and copy the `id` from the browser URL. The ID should look like `b783bf32-0cc2-4d31-aadc-22d4c4807c30`.

Set the `id` field in the `d2.config.js` in your app to the id of the app on the App Hub.

The `minDHIS2Version` is compulsory as well, so set the minimum DHIS2 version that this app is compatible with.

> ```

Reviewer: ```suggestion
```js title="d2.config.js"
```

> // d2.config.js

Reviewer: ```suggestion
 
```
